### PR TITLE
feat(toast): add global config toastDuration

### PR DESCRIPTION
**Pull request by DwieDima**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

- there if no option to define toast duration globally

Issue URL: [#22497](https://github.com/ionic-team/ionic-framework/issues/22497)

## What is the new behavior?

- define toast duration globally

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by liamdebeasi** [comment]
Thanks for the PR! The team is working on a design document for this feature, so we will review your PR when that is complete.

File: core/src/components/toast/toast.tsx [hunk 1]
```diff
@@ -77,7 +78,7 @@ export class Toast implements ComponentInterface, OverlayInterface {
    * How many milliseconds to wait before hiding the toast. By default, it will show
    * until `dismiss()` is called.
    */
-  @Prop() duration = 0;
+  @Prop() duration = config.get('toastDuration', 0);
```
Comment (liamdebeasi) [inline core/src/components/toast/toast.tsx:81]: Can we use `config.getNumber`? This method has typing to return `number` which should also remove the `component.d.ts` diff.

Comment (liamdebeasi) [inline core/src/components/toast/toast.tsx:81]: Can we also add a test for this? You could write a test that sets `window.Ionic = { config: { toastDuration: 5000 } }` and checks that the `duration` property on `ion-toast` defaults to `5000`. We should also check to make sure that it defaults to `0` when the config option is not set.

File: core/src/utils/config.ts [hunk 2]
```diff
@@ -96,6 +96,11 @@ export interface IonicConfig {
    */
   tabButtonLayout?: TabButtonLayout;
 
+  /**
+   * Overrides the default duration for all `ion-toast` overlays.
```
Comment (liamdebeasi) [inline core/src/utils/config.ts:100]:
```suggestion
   * Overrides the default `duration` for all `ion-toast` components.
```

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
These additions look great. A few small changes we would like to make.

File: core/src/components/toast/toast.tsx (hunk 1 above)
Comment (DwieDima) [reply core/src/components/toast/toast.tsx]: Yes of course. Added 2 tests.

File: core/src/components/toast/test/basic/toast.e2e.ts [hunk 3]
```diff
@@ -133,4 +133,15 @@ test.describe('toast: properties', () => {
...
     await expect(toast).toHaveClass(/my-custom-class/);
   });
+
+  test('should have duration set to 0', async () => {
+    const { toast } = await toastFixture.openToast('#show-bottom-toast');
+    await expect(toast).toHaveAttribute('duration', '0');
+  });
+
+  test('should have duration set to 5000', async () => {
+    window.Ionic = { config: { toastDuration: 5000 } };
```
[the remark this answers is not in the document]
Comment (DwieDima) [reply core/src/components/toast/test/basic/toast.e2e.ts]: ok, thank you very much for the tip!

File: core/src/components/toast/test/basic/toast.e2e.ts [hunk 4]
```diff
@@ -134,3 +134,27 @@ test.describe('toast: properties', () => {
     await expect(toast).toHaveClass(/my-custom-class/);
   });
 });
+
+test.describe('toast: duration config', () => {
+  test('should have duration set to 0', async ({ page }) => {
```
Comment (liamdebeasi) [inline core/src/components/toast/test/basic/toast.e2e.ts:139]:
```suggestion
  test.beforeEach(({ skip }) => {
    skip.rtl();
    skip.mode('ios');
  });
  test('should have duration set to 0', async ({ page }) => {
```

One last change -- we don't need to run this for RTL/multi mode layouts.

**Review by liamdebeasi — APPROVED** [review]
Changes look good. Thank you!

File: core/src/components/toast/test/basic/toast.e2e.ts [hunk 5]
```diff
@@ -134,3 +134,31 @@ test.describe('toast: properties', () => {
...
+    <ion-toast></ion-toast>
+    <script>
+      window.Ionic = {
+        config: { toastDuration: 5000 } 
+      }
+    </script>
+  `);
+
+    const toast = page.locator('ion-toast');
+    await expect(toast).toHaveAttribute('duration', '5000');
```
Comment (liamdebeasi) [inline core/src/components/toast/test/basic/toast.e2e.ts:162]: Sorry, I mistyped before. This needs to be `toHaveJSProperty` not `toHaveAttribute`

Comment (liamdebeasi) [inline core/src/components/toast/test/basic/toast.e2e.ts:162]:
```suggestion
    await expect(toast).toHaveJSProperty('duration', 5000);
```

Sorry, I mistyped before. This needs to be `toHaveJSProperty` not `toHaveAttribute`

File: core/src/components/toast/test/basic/toast.e2e.ts [hunk 6]
```diff
@@ -134,3 +134,31 @@ test.describe('toast: properties', () => {
...
+  test.beforeEach(({ skip }) => {
+    skip.rtl();
+    skip.mode('ios');
+  });
+  test('should have duration set to 0', async ({ page }) => {
+    await page.setContent(`
+    <ion-toast></ion-toast>
+  `);
+    const toast = page.locator('ion-toast');
+    await expect(toast).toHaveAttribute('duration', '0');
```
Comment (liamdebeasi) [inline core/src/components/toast/test/basic/toast.e2e.ts:148]:
```suggestion
    await expect(toast).toHaveJSProperty('duration', 0);
```

Same issue here

**Review by sean-perkins — APPROVED** [review]

**Review by amandaejohnston — APPROVED** [review]

**Comment by amandaejohnston** [comment]
Merged in, thank you very much for your contribution!

**Pull request merged**